meta_review/handler.py

```diff
@@ -0,0 +1,412 @@
+from dateutil import parser
+import json
+import logging
+import os
+
+from django.utils import timezone
+
+from meta_review.dump_to_db import dump_all
+from meta_review.load_from_db import load_all
+from meta_review.models import Participant
+
+
+def parse_time(time):
+    """
+    parse string to datetime
+    :param time: a string represents time, e.g. 2018-05-09T11:19:26Z
+    :return: an offset-aware datetime object
+    """
+    if time is None:
+        return None
+
+    return parser.parse(time)
+
+
+class MetaReviewHandler:
+    """
+    This is the class responsible for scraping provided information (reviews,
+    reactions), processing them and dumping into Django database
+    """
+
+    def __init__(self, content, date):
+        """
+        Constructs a new ``MetaReviewHandler``
+
+        :param content: Parsed JSON data
+        :param date: The update date
+        """
+        self.logger = logging.getLogger(__name__)
+        self.logger.info('this package is alive')
+
+        self.date = date
+
+        # save comments in memory
+        self.comments = dict()
+        for issue in content:
+            issue = issue['issue']
+            if not issue.get('pullRequest'):
+                continue
+            for comment in issue['pullRequest']['comments']:
+                # parse time
+                comment['createdAt'] = parse_time(comment['createdAt'])
+                comment['lastEditedAt'] = parse_time(comment['lastEditedAt'])
+                self.comments[comment['id']] = comment
+
+        # save reactions in memory
+        self.reactions = dict()
+        for key, comment in self.comments.items():
+            if not comment.get('reactions'):
+                continue
+            for reaction in comment['reactions']:
+                # record receiver
+                reaction['receiver'] = {
+                    'login': comment['author']['login']
+                }
+                # record comment id
+                reaction['comment_id'] = comment['id']
+                # parse time
+                reaction['createdAt'] = parse_time(reaction['createdAt'])
+                self.reactions[reaction['id']] = reaction
+
+        # save participants in memory
+        self.participants = dict()
+        for key, comment in self.comments.items():
+            # get author of each comment
+            author = comment['author']
+
+            # skip if user does not exist
+            # this happens when account is deleted from GitHub
+            if author['login']:
+                self.participants[author['login']] = author
+
+            if not comment.get('reactions'):
+                continue
+            for reaction in comment['reactions']:
+                # get user of each reaction
+                user = reaction['user']
+
+                # skip if user does not exist
+                if author['login']:
```

Comment:
should this be `user['login']` ?

Reply:
Yeah, a bug here. This would lead to an exception if the user of reaction does not exist. Either GitHub already handled this, or we just didn't encounter it.